alibi_detect/od/pytorch/ensemble.py

```diff
@@ -0,0 +1,341 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from typing import Optional
+
+import torch
+import numpy as np
+from torch.nn import Module
+
+from alibi_detect.base import NotFitException
+
+
+class BaseTransformTorch(Module, ABC):
+    def __init__(self):
+        """Base Transform class.
+
+        provides abstract methods for transform objects that map `torch` tensors.
+        """
+        super().__init__()
+
+    def transform(self, x: torch.Tensor):
+        return self._transform(x)
+
+    @abstractmethod
+    def _transform(self, x: torch.Tensor):
+        """Applies class transform to `torch.Tensor`
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` array to be transformed
+        """
+        pass
+
+    @torch.no_grad()
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.transform(x=x)
+
+
+class FitMixinTorch(ABC):
+    fitted = False
+
+    def __init__(self):
+        """Fit mixin
+
+        Utility class that provides fitted checks for alibi-detect objects that require to be fit before use.
+        """
+        super().__init__()
+
+    def fit(self, x: torch.Tensor) -> FitMixinTorch:
+        self.fitted = True
+        self._fit(x)
+        return self
+
+    @abstractmethod
+    def _fit(self, x: torch.Tensor):
+        """Fit on `x` tensor.
+
+        This method should be overidden on child classes.
+
+        Parameters
+        ----------
+        x
+            Reference `torch.Tensor` for fitting object.
+        """
+        pass
+
+    @torch.jit.unused
+    def check_fitted(self):
+        """Raises error if parent object instance has not been fit.
+
+        Raises
+        ------
+        NotFitException
+            Raised if method called and object has not been fit.
+        """
+        if not self.fitted:
+            raise NotFitException(f'{self.__class__.__name__} has not been fit!')
+
+
+class BaseFittedTransformTorch(BaseTransformTorch, FitMixinTorch):
+    def __init__(self):
+        """Base Fitted Transform class.
+
+        Extends `BaseTransfrom` with fit functionality. Ensures that transform has been fit prior to
+        applying transform.
+        """
+        BaseTransformTorch.__init__(self)
+        FitMixinTorch.__init__(self)
```

Comment:
Think we can just do `super().__init__()` even with multiple inheritence, plus, as mentioned above, a mixing shouldn't need a constructor.